Mathlib/Topology/ProperMap.lean

```diff
@@ -308,37 +315,95 @@ lemma isProperMap_iff_isClosedMap_and_tendsto_cofinite [T1Space Y] :
+theorem compactlyGenerated_of_sequentialSpace [T2Space X] [SequentialSpace X] {s : Set X} :
+    IsClosed s ↔ ∀ ⦃K⦄, IsCompact K → IsClosed (s ∩ K) := by
+  refine ⟨fun hs K hK ↦ hs.inter hK.isClosed,
+    fun h ↦ SequentialSpace.isClosed_of_seq _ fun u p hu hup ↦
+    mem_of_mem_inter_left ((h hup.isCompact_insert_range).mem_of_tendsto hup ?_)⟩
+  simp only [mem_inter_iff, mem_insert_iff, mem_range, exists_apply_eq_apply, or_true, and_true,
+    eventually_atTop, ge_iff_le]
+  exact ⟨0, fun n _ ↦ hu n⟩
+
+/-- If `Y` is sequential and Hausdorff, then proper maps `X → Y` are exactly continuous maps
+such that the preimage of any compact set is compact.
+
+This result is a direct consequence of `isProperMap_iff_isCompact_preimage`, because any
+Hausdorff and sequential space is compactly generated. In the future it should be inferred
+by typeclass inference, however compactly generated spaces are not yet in Mathlib, therefore
+we also add this theorem. -/
+theorem SequentialSpace.isProperMap_iff_isCompact_preimage [T2Space Y] [SequentialSpace Y] :
+    IsProperMap f ↔ Continuous f ∧ ∀ ⦃K⦄, IsCompact K → IsCompact (f ⁻¹' K) :=
+  _root_.isProperMap_iff_isCompact_preimage
+    (fun _ ↦ compactlyGenerated_of_sequentialSpace)
 
 /-- Version of `isProperMap_iff_isCompact_preimage` in terms of `cocompact`. -/
-lemma isProperMap_iff_tendsto_cocompact [T2Space Y] [WeaklyLocallyCompactSpace Y] :
+lemma isProperMap_iff_tendsto_cocompact [T2Space Y]
+    (compactlyGenerated : ∀ s : Set Y, IsClosed s ↔ ∀ ⦃K⦄, IsCompact K → IsClosed (s ∩ K)) :
     IsProperMap f ↔ Continuous f ∧ Tendsto f (cocompact X) (cocompact Y) := by
-  simp_rw [isProperMap_iff_isCompact_preimage, hasBasis_cocompact.tendsto_right_iff,
-    ← mem_preimage, eventually_mem_set, preimage_compl]
+  simp_rw [isProperMap_iff_isCompact_preimage compactlyGenerated,
+    hasBasis_cocompact.tendsto_right_iff, ← mem_preimage, eventually_mem_set, preimage_compl]
   refine and_congr_right fun f_cont ↦
     ⟨fun H K hK ↦ (H hK).compl_mem_cocompact, fun H K hK ↦ ?_⟩
   rcases mem_cocompact.mp (H K hK) with ⟨K', hK', hK'y⟩
   exact hK'.of_isClosed_subset (hK.isClosed.preimage f_cont)
     (compl_le_compl_iff_le.mp hK'y)
 
+/-- Version of `WeaklyLocallyCompactSpace.isProperMap_iff_isCompact_preimage`
+in terms of `cocompact`. -/
+lemma WeaklyLocallyCompactSpace.isProperMap_iff_tendsto_cocompact [T2Space Y]
+    [WeaklyLocallyCompactSpace Y] :
+    IsProperMap f ↔ Continuous f ∧ Tendsto f (cocompact X) (cocompact Y) :=
+  _root_.isProperMap_iff_tendsto_cocompact
+    (fun _ ↦ compactlyGenerated_of_weaklyLocallyCompactSpace)
+
+/-- Version of `SequentialSpace.isProperMap_iff_isCompact_preimage` in terms of `cocompact`. -/
+lemma SequentialSpace.isProperMap_iff_tendsto_cocompact [T2Space Y] [SequentialSpace Y] :
+    IsProperMap f ↔ Continuous f ∧ Tendsto f (cocompact X) (cocompact Y) :=
+  _root_.isProperMap_iff_tendsto_cocompact
+    (fun _ ↦ compactlyGenerated_of_sequentialSpace)
```

Comment:
Sounds like you should define the typeclass `Topology.IsCompactlyGenerated`

Reply:
The idea of generalizing `isProperMap_iff_isCompact_preimage` to compactly generated spaces comes from @ADedecker, but he told me that they were hesitating on the definition to choose (they are many different possible, which are not equivalent). I picked one but I feel like I shouldn't be the one to choose the Mathlib definition.

Reply:
I'll let Anatole decide here

Reply:
Yes the point was that we have to choose the best definition among the three defined on [Wikipedia](https://en.wikipedia.org/wiki/Compactly_generated_space). I think #13452 would allow to formulate definition 1 and 3, but probably not option 2. I meant to take the time to study the question a bit more, but I didn't get to do it. I suggested this ad-hoc approach to Étienne because it contains all the mathematics so switching to a more typical setup would be straightforward.

Reply:
I think that the docstring should discuss this.